src/OrchardCore.Modules/OrchardCore.HealthChecks/Startup.cs

```diff
@@ -17,17 +21,38 @@ public Startup(IShellConfiguration shellConfiguration)
             _shellConfiguration = shellConfiguration;
         }
 
-        public override void Configure(IApplicationBuilder app, IEndpointRouteBuilder routes, IServiceProvider serviceProvider)
+        public override void ConfigureServices(IServiceCollection services)
         {
-            var healthCheckOptions = serviceProvider.GetService<IOptions<HealthChecksOptions>>().Value;
+            services.AddScoped<IHealthChecksResponseWriter, DefaultHealthChecksResponseWriter>();
+            services.AddHealthChecks();
 
-            routes.MapHealthChecks(healthCheckOptions.Url);
+            services.Configure<HealthChecksOptions>(_shellConfiguration.GetSection("OrchardCore_HealthChecks"));
         }
 
-        public override void ConfigureServices(IServiceCollection services)
+        public override void Configure(IApplicationBuilder app, IEndpointRouteBuilder routes, IServiceProvider serviceProvider)
         {
-            services.AddHealthChecks();
-            services.Configure<HealthChecksOptions>(_shellConfiguration.GetSection("OrchardCore_HealthChecks"));
+            var healthChecksOptions = serviceProvider.GetService<IOptions<HealthChecksOptions>>().Value;
+
+            if (healthChecksOptions.ShowDetails)
+            {
+                var healthChecksResponseWriter = serviceProvider.GetService<IHealthChecksResponseWriter>();
+
+                app.UseHealthChecks(healthChecksOptions.Url, new HealthCheckOptions
+                {
+                    AllowCachingResponses = false,
+                    ResultStatusCodes =
+                    {
+                        [HealthStatus.Healthy] = StatusCodes.Status200OK,
+                        [HealthStatus.Degraded] = StatusCodes.Status500InternalServerError,
+                        [HealthStatus.Unhealthy] = StatusCodes.Status503ServiceUnavailable
+                    },
```

Comment:
Why override the default?

Reply:
Because `HealthStatus.Degraded` was shows `200` by default

Reply:
Yeah, but why is that a problem?

Reply:
I think degraded is sort of problematic that need a proper status code

Reply:
Yeah, but I guess the ASP.NET Core developers intentionally didn't have that either. My problem here is, that the health check endpoint is an API in itself, that communicates about the health of the app. HTTP 5xx means that something is broken on the server for the current request; in a non-healthy state that may be true, but really only when the health check endpoint itself is broken.

So, I'd actually go in the other direction, and always return a 2xx (looking at those status codes, the only one fitting I see is 200, i.e. return 200 for all states). This only means that the health check endpoint works. Then, the response body will tell whether the app is healthy (after all, this is why we have a body) or not. If there's anything else, than that means that health check is also broken, and the app is definitely not healthy.

Please research whether this has any drawbacks why we shouldn't do it.

Reply:
503 "Service Unavailable" makes sense to me because the specific service is broken, I don't think 200 suits all. We could use 200 for degraded as shown in the [link](https://learn.microsoft.com/en-us/aspnet/core/host-and-deploy/health-checks?view=aspnetcore-8.0#customize-the-http-status-code)

Reply:
But the service is not necessarily broken. There can be very well cases where the health check works, but the rest of the site is down. Then the health check can correctly return the status (which is degraded/unhealthy), with itself working, with a 200. Then any non-2xx status could would mean that health check itself is broken too.

How would you otherwise distinguish between health check being broken and the rest of the app working, vs the app being broken and health check working? Because it can also very well be that the app is healthy, but we messed up something with the health check providers.

Reply:
This isn't very clear :) assume X mail service is broken, we know the app is healthy but the service itself is broken, so the response above is related to the health check itself, not the app

Reply:
In that case, the response body of the health check will elaborate that Email is broken. I don't see why responding with an HTTP 500 in addition to that would help.

The only use-case I can think is when you ping the health check endpoint with some tool that can't read the response but can only check for an HTTP 200 response code. However, that's really a shortcoming of that tool, then, and you should use something more capable for health checks (and only use such simple pings for checking e.g. the homepage).

Reply:
If you are against the health result status codes, I suggest you open an issue in the AspNetCore repo, and then the team will answer why they chose these codes

Reply:
I'm not against having that capability. But the ASP.NET team gave us the option to specify which status code to use, what you also did here. So, we can make our own decisions on those status codes, just as you did. I'm arguing for using a different configuration than what you opted for.

Reply:
So, let rely on the defaults then we can extend and make these codes configurable

Reply:
I've presented my arguments above. I really don't want to have to repeat those at a later time.

Reply:
I know, but I'm not convinced with your arguments :)

Reply:
I welcome you to please present your counter-arguments, then.

Reply:
> but really only when the health check endpoint itself is broken

Not necessary, if you look to https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/503 indicates the service is not ready to handle the request, which is true in terms of health check service, so the service is unhealthy and might require you to retry later

While the health check reports the health of app infrastructure components, how unhealthy components should return 200 if something goes wrong behind the scenes, I know someone will tell me the endpoint is not broken but in reality it's unavailable IMHO

Reply:
Would you say that if the health check endpoint is fully working and returns a correct JSON that describes the app otherwise being down, that then the endpoint itself is "Unavailable"?

Reply:
Take a GitHub as an example https://www.githubstatus.com/ if `Codespaces` for instance is down for any reason this component is `Unhealthy`, so in terms of health checks the entire app is unhealthy but that doesn't mean  the whole app is down

Reply:
Yep, and that page follows what I think is a better approach: If Codespaces is down, or even if the whole of GitHub is down, https://www.githubstatus.com/ will still return HTTP 200, and report issues with the services. The only case when https://www.githubstatus.com/ will return 5xx is when https://www.githubstatus.com/ itself (not the rest of GitHub) is broken. The health check endpoint is the same thing, just a simpler approach than a separate status page.

Note that I'm not talking about the Health/Unhealthy flag; I don't have any issues with what's in this PR for that (and I agree that "if Codespaces is down then GitHub as a whole is Unhealthy"). Rather, the only thing I'm talking about is the HTTP status code of the health check endpoint (i.e. what I added the PR comment on above), which I think should be 200 even if the service is unhealthy.

Reply:
I revised the PR and the ASP.NET Core docs, both `Healthy` and `Degraded` should return 200, but for `Unhealthy` 503 is the better option I found and agree with. Service is Unavailable makes sense

So let us merge this, and react to the feedback at the next meeting

Reply:
I elaborated why it's not better and I'm awaiting your counter arguments on my last comment. There's no point in not settling this now and I absolutely don't want either of us to have to get back to this and kicking the can down the road.

Reply:
Also, I elaborate as much as I can on why unhealthy service should return 503 not 200 because the service is unavailable for some reason, I'm not sure why a useless or damaged service should return 200, that's why the default

Let's rely on the default status or hear the feedback from other guys if you never mind, so that PR will not take much time

Reply:
Last example that might change your mind :)

If we have 3 health check services, according to your assumption if Service 1 is unhealthy the status will be 200 for the health reporter, as a user will be happy because the health check reported is 200 while it's not.

If there's no "ShowDetails" option that I have added is there you can recognize if there's an unhealthy service, that's why each health check service should respond either 200 or 503 to notify the reporter if the entire system is healthy or not. Of course, the app SHOULD return the status based on all the health check services

I know this takes so long but I tried in different ways to convince you that we should differentiate between the end-point existence and health check status

Reply:
Is there really a use case when we don't want the health check endpoint to respond with details, and the different services to not provide details when they report themselves being unhealthy?

Reply:
Nope, but responding with 200 does not make sense to me even if you said the API hits :) but that does not mean the response is OK

Let me give you a last example assume you have an API to create a post, for any reason during the post creating a system throws an expectation SHOULD the API respond with 200 or 500?

Reply:
We shouldn't allow empty health check responses, then, and thus remove configuration that we want to discourage (`ShowDetails`).

It should respond with 500. But this is precisely my point :). If the API for creating a blog post is broken, then it should indicate that with a 5xx status code. But you wouldn't want the blog post API return a 500 if only creating a page is broken, right? Only when you want to create a page. Because the page API being broken doesn't mean that the blog API is also broken.

If the health check endpoint is broken then it should return a 5xx too. However, I'm talking about the cases when the health check endpoint itself is working fine, but the rest of the app is unhealthy (then it should still stay 200).

We're probably taking waaaaay too much on this issue, but let me illustrate it with something else. Suppose you see this sign at the side of the road:

![image](https://github.com/OrchardCMS/OrchardCore/assets/1976647/c9da491a-874d-4cb5-8b81-37ddd5e50a27)

The sign is working perfectly, but it's telling you about heat and fire. Looks good, right? :) To me, returning 500 is akin to making sure this sign is on fire too, or at least it says "I'm also on fire!". But we don't do that, because we can (and should) distinguish between the sign being broken, and the thing that the sign tells about being broken, because those are two different things.

Reply:
> We shouldn't allow empty health check responses

Please elaborate.

> However, I'm talking about the cases when the health check endpoint itself is working fine, but the rest of the app is unhealthy (then it should still stay 200).

I don't think so :) because it tells me that the entire app is healthy, which is not. For that, I need anyone else on the team to share his opinion, maybe I'm wrong

Reply:
> We shouldn't allow empty health check responses

[You said so](https://github.com/OrchardCMS/OrchardCore/pull/13590#discussion_r1565006360) :).

Why would it say the entire app is healthy? The health check will literally return `"unhealthy"`. HTTP 200 only means that this response was successful.

Reply:
No, not at all, this is not my point. It's that the status of the health check API, the thing that tells about the state of the app, is not the same as the status of the app. The road sign can be fine when it tells that there's a fire ahead. Or githubstatus.com can be fine while github.com, the service it reports about, is broken.